src/neural/network_cudnn.cu

```diff
@@ -1001,7 +1000,53 @@ class CudnnNetwork : public Network {
 
     // 1. allocate scratch space (used internally by cudnn to run convolutions,
     //     and also for format/layout conversion for weights)
-    reportCUDAErrors(cudaMalloc(&scratch_mem_, kCudaScratchSize));
+    cudnnFilterDescriptor_t wDesc;
+    cudnnConvolutionDescriptor_t convDesc;
+    cudnnTensorDescriptor_t xDesc;
+    cudnnCreateFilterDescriptor(&wDesc);
+    cudnnCreateConvolutionDescriptor(&convDesc);
+    cudnnCreateTensorDescriptor(&xDesc);
+    cudnnConvolutionFwdAlgo_t convAlgo;
+
+    const int maxChannels =
+        kInputPlanes > numFilters ? kInputPlanes : numFilters;
+
+    const bool fp16 = std::is_same<half, DataType>::value;
+    reportCUDNNErrors(cudnnSetFilter4dDescriptor(
+        wDesc, fp16 ? CUDNN_DATA_HALF : CUDNN_DATA_FLOAT,
+        fp16 ? CUDNN_TENSOR_NHWC : CUDNN_TENSOR_NCHW, maxChannels, maxChannels,
+        3, 3));
+
+    reportCUDNNErrors(cudnnSetTensor4dDescriptor(
+        xDesc, fp16 ? CUDNN_TENSOR_NHWC : CUDNN_TENSOR_NCHW,
+        fp16 ? CUDNN_DATA_HALF : CUDNN_DATA_FLOAT, kMaxBatchSize, maxChannels,
+        8, 8));
+
+    reportCUDNNErrors(cudnnSetConvolution2dDescriptor(
+        convDesc, 1, 1, 1, 1, 1, 1, CUDNN_CROSS_CORRELATION,
+        fp16 ? CUDNN_DATA_HALF : CUDNN_DATA_FLOAT));
+
+    if (fp16)
+      reportCUDNNErrors(
+          cudnnSetConvolutionMathType(convDesc, CUDNN_TENSOR_OP_MATH));
+
+    if (!fp16)
```

Comment:
Maybe replace `!fp16` with `fp16` and swap "if" and "else" statements.
Also consider merging with previous "if".